keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -887,19 +894,19 @@ routed according to the families assigned for services.
 
 1. For Headless with selector, the value of `spec.ipFamilies` will drive the
    endpoint selection. Endpoints will use the value of `ipFamiles[0]`.
-   EndpointSlices will use endpoints with all IP families of the Service.
-2. For Headless without selector the values of `spec.preferDualStack` and
-   `spec.ipFamilies` are defaulted and saved as normal, but are ignored by
-   kubernetes (since the standard endpoint controllers ignore such services).
-
-> In all cases, the values of `spec.preferDualStack`, `spec.ipFamilies`, and
-> `spec.clusterIPs` are validated and must be consistent.
+   EndpointSlices will use endpoints with all IP families of the Service. This
+   value is expected to match cluster configuration. i.e. creating  service with
+   a family not configured on server will generate an error.
+2. For Headless without selector the values of `spec.ipFamilyPolicy` and
+   `spec.ipFamilies` are semantically validated for correctness but are not 
+   validated against cluster configuration. This means a headless service without
+   selectors can have IP families that are not configured on server.
 
 #### Type ExternalName
 
-For ExternalName service the value of `spec.preferDualStack` is ignored and
-`spec.ipFamilies` is expected to be `nil`. Values will be set to `nil` by the
-apiserver.
+For ExternalName service the value of `spec.ipFamilyPolicy` and
+`spec.ipFamilies` is expected to be `nil`. Any other value for these fields 
+will cause validation errors.
```

Comment:
should specify that when converting from ClusterIP to ExternalName, those fields get cleared automatically

Reply:
i disagree on that. when converting user should set them nil. Same rule apply. intent has to be clear.

Reply:
That would be an API break; if the user is not intentionally making use of new functionality, we can't require that they be aware of newly-introduced fields. In 1.18, you can create a `type: ClusterIP` service, then patch it to change `type` to `ExternalName` and not make any other changes, and that will work. So that has to continue working.

Reply:
The patch has to clear `clusterIP` manually, but I think Dan's right.  We can't now expand that to require the user clear the new fields.  @liggitt is adding a mechanism to pass back warnings on API calls.  Jordan, do you think "I auto-cleared this other field" would fall under that?

Reply:
> Jordan, do you think "I auto-cleared this other field" would fall under that?

I'm not sure. My original thinking was to limit warnings to cases where a user was sending API requests they should change (duplicate fields, known bad values, etc) or stop sending (deprecated API use, etc).

Sending a warning for an API request that was completely correct in a previous version and which is not deprecated seems not great.

Reply:
Yeah, I buy that.  It felt like a stretch.  We're going to have to open this can of worms (if it isn't already) which allows mutations (specifically clearing) of fields that may not have been specified in a transaction in order to preserve internal consistency.

For example, the need to manually clear NodePort fields when change service types is a constant pain in the butt.